integration/v3_transaction_test.go

```diff
@@ -0,0 +1,319 @@
+//go:build rpctest
+// +build rpctest
+
+package integration
+
+import (
+	"testing"
+
+	"github.com/btcsuite/btcd/chaincfg"
+	"github.com/btcsuite/btcd/integration/rpctest"
+	"github.com/btcsuite/btcd/txscript"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/stretchr/testify/require"
+)
+
+// TestV3TransactionPolicy tests that v3 transactions are accepted by the
+// mempool and can be mined into blocks.
+func TestV3TransactionPolicy(t *testing.T) {
+	t.Parallel()
+
+	// Setup the test harness, and initialize the harness with mature
+	// coinbase outputs for spending.
+	btcdCfg := []string{"--rejectnonstd", "--debuglevel=debug"}
+	r, err := rpctest.New(&chaincfg.SimNetParams, nil, btcdCfg, "")
+	require.NoError(t, err)
+
+	require.NoError(t, r.SetUp(true, 100))
+	t.Cleanup(func() {
+		require.NoError(t, r.TearDown())
+	})
+
+	// Make sure that normal v1 transactions are still accepted.
+	t.Run("v1 transaction accepted", func(t *testing.T) {
+		testVersionedTransaction(t, r, 1, true)
+	})
+
+	// MAke sure that v2 transactions are still accepted.
+	t.Run("v2 transaction accepted", func(t *testing.T) {
+		testVersionedTransaction(t, r, 2, true)
+	})
+
+	// Ensure that v3 transactions are accepted.
+	t.Run("v3 transaction accepted", func(t *testing.T) {
+		testVersionedTransaction(t, r, 3, true)
+	})
+
+	// Ensure that v4 transactions are rejected.
+	t.Run("v4 transaction rejected", func(t *testing.T) {
+		testVersionedTransaction(t, r, 4, false)
+	})
+
+	// Test that v3 transactions can be mined.
+	t.Run("v3 transaction mined in block", func(t *testing.T) {
+		testV3TransactionMining(t, r)
+	})
+}
+
+// testVersionedTransaction creates a transaction with the specified version
+// and tests whether it's accepted by the mempool.
+func testVersionedTransaction(t *testing.T, r *rpctest.Harness, version int32,
+	shouldBeAccepted bool) {
+
+	addr, err := r.NewAddress()
+	require.NoError(t, err)
+
+	script, err := txscript.PayToAddrScript(addr)
+	require.NoError(t, err)
+
+	output := &wire.TxOut{
+		PkScript: script,
+		Value:    50000000,
+	}
+
+	var tx *wire.MsgTx
+	if version == 1 {
```

Comment:
Super nitpicking but can't you just replace the entire if branching with:

`r.CreateTransaction([]*wire.TxOut{output}, 10, true, rpctest.WithTxVersion(version))`.

Not sure why there needs to be a check for `version == 1`.

Reply:
Totally right, simplified.